File: pricing/03-price-comparison/price_comparison.ts
/**
 * Price comparison across HarchOS regions and tiers (TypeScript).
 *
 * Compares GPU pricing plans across different regions, tiers, and GPU types
 * to find the most cost-effective deployment. Optionally factors in carbon
 * intensity for carbon-adjusted cost scoring.
 *
 * Usage:
 *   npx ts-node price_comparison.ts
 *   npx ts-node price_comparison.ts --gpu-type H100
 *   npx ts-node price_comparison.ts --include-carbon
 */

import { HarchOSClient } from "harchos";
import type { PricingPlan, ListPricingPlansParams } from "harchos";

// ---------------------------------------------------------------------------
// Simulated data
// ---------------------------------------------------------------------------

interface SimPlan extends PricingPlan {
  hub?: string;
  carbonIntensity?: number;
  renewablePct?: number;
}

const SIMULATED_PLANS: SimPlan[] = [
  { id: "p1", name: "H100 Enterprise (Ouarzazate)", gpuType: "H100",
    pricePerGpuHour: 2.10, pricePerCpuCoreHour: 0.04, pricePerGbStorageMonth: 0.08,
    pricePerGbMemoryHour: 0.005, currency: "USD", region: "morocco",
    tier: "enterprise", isDefault: true, hub: "Ouarzazate", carbonIntensity: 18, renewablePct: 97.2 },
  { id: "p3", name: "H100 Performance (Benguerir)", gpuType: "H100",
    pricePerGpuHour: 2.35, pricePerCpuCoreHour: 0.045, pricePerGbStorageMonth: 0.09,
    pricePerGbMemoryHour: 0.005, currency: "USD", region: "morocco",
    tier: "enterprise", isDefault: false, hub: "Benguerir", carbonIntensity: 55, renewablePct: 88.5 },
  { id: "p4", name: "A100 Performance (Tanger)", gpuType: "A100",
    pricePerGpuHour: 1.80, pricePerCpuCoreHour: 0.035, pricePerGbStorageMonth: 0.07,
    pricePerGbMemoryHour: 0.004, currency: "USD", region: "morocco",
    tier: "enterprise", isDefault: false, hub: "Tanger", carbonIntensity: 95, renewablePct: 82.1 },
  { id: "p5", name: "A100 Standard (Casablanca)", gpuType: "A100",
    pricePerGpuHour: 1.95, pricePerCpuCoreHour: 0.038, pricePerGbStorageMonth: 0.075,
    pricePerGbMemoryHour: 0.0045, currency: "USD", region: "morocco",
    tier: "community", isDefault: false, hub: "Casablanca", carbonIntensity: 210, renewablePct: 45.0 },
  { id: "p6", name: "L40S Enterprise (Dakhla)", gpuType: "L40S",
    pricePerGpuHour: 1.40, pricePerCpuCoreHour: 0.03, pricePerGbStorageMonth: 0.06,
    pricePerGbMemoryHour: 0.003, currency: "USD", region: "morocco",
    tier: "enterprise", isDefault: false, hub: "Dakhla", carbonIntensity: 32, renewablePct: 94.8 },
  { id: "p7", name: "L40S Performance (Benguerir)", gpuType: "L40S",
    pricePerGpuHour: 1.55, pricePerCpuCoreHour: 0.032, pricePerGbStorageMonth: 0.065,
    pricePerGbMemoryHour: 0.0035, currency: "USD", region: "morocco",
    tier: "enterprise" as any, isDefault: false, hub: "Benguerir", carbonIntensity: 55, renewablePct: 88.5 },
];

// Carbon cost: $50/tonne CO2
const CARBON_COST_PER_KG = 0.05;

// ---------------------------------------------------------------------------
// Comparison logic
// ---------------------------------------------------------------------------

interface ComparisonResult {
  id: string;
  name: string;
  gpuType: string;
  tier: string;
  costPerGpuHour: number;
  monthlyCost: number;
  currency: string;
  carbonIntensity?: number;
  renewablePct?: number;
  co2PerGpuHrKg?: number;
  carbonCostPerHr?: number;
  effectiveCostPerHr?: number;
  effectiveMonthlyCost?: number;
}

function comparePrices(
  plans: SimPlan[],
  hours: number,
  gpuCount: number,
  includeCarbon: boolean,
): ComparisonResult[] {
  const usdPlans = plans.filter((p) => p.currency === "USD");

  const results: ComparisonResult[] = usdPlans.map((plan) => {
    const monthlyCost = plan.pricePerGpuHour * gpuCount * hours;
    const result: ComparisonResult = {
      id: plan.id,
      name: plan.name,
      gpuType: plan.gpuType,
      tier: plan.tier,
      costPerGpuHour: plan.pricePerGpuHour,
      monthlyCost: Math.round(monthlyCost * 100) / 100,
      currency: plan.currency,
    };

    if (includeCarbon) {
      const carbonIntensity = plan.carbonIntensity || 100;
      const co2PerGpuHr = (0.3 * carbonIntensity) / 1000;
      const carbonCostPerHr = co2PerGpuHr * CARBON_COST_PER_KG;
      const effectiveCostPerHr = plan.pricePerGpuHour + carbonCostPerHr;

      result.carbonIntensity = carbonIntensity;
      result.renewablePct = plan.renewablePct || 0;
      result.co2PerGpuHrKg = Math.round(co2PerGpuHr * 10000) / 10000;
      result.carbonCostPerHr = Math.round(carbonCostPerHr * 10000) / 10000;
      result.effectiveCostPerHr = Math.round(effectiveCostPerHr * 10000) / 10000;
      result.effectiveMonthlyCost = Math.round(effectiveCostPerHr * gpuCount * hours * 100) / 100;
    }

    return result;
  });

  const sortKey = includeCarbon ? "effectiveCostPerHr" : "costPerGpuHour";
  results.sort((a, b) => (a[sortKey] || a.costPerGpuHour) - (b[sortKey] || b.costPerGpuHour));

  return results;
}

function printComparison(results: ComparisonResult[], includeCarbon: boolean): void {
  if (!results.length) {
    console.log("\n  No plans found.");
    return;
  }

  const medals = ["🥇", "🥈", "🥉"];

  console.log(
    `\n  ${"Rank".padStart(4)} | ${"Plan Name".padEnd(38)} | ${"GPU".padEnd(6)} | ` +
    `${"Tier".padEnd(12)} | ${"$/gpu/hr".padStart(8)} | ${"Monthly (1 GPU)".padStart(15)}` +
    (includeCarbon
      ? ` | ${"CO2 kg/hr".padStart(10)} | ${"Eff. $/hr".padStart(9)} | ${"Renewable".padStart(9)}`
      : "")
  );
  console.log("  " + "-".repeat(includeCarbon ? 140 : 100));

  for (let i = 0; i < results.length; i++) {
    const r = results[i];
    const medal = medals[i] || "  ";
    let line =
      `  ${medal} ${(i + 1).toString().padStart(2)} | ${r.name.padEnd(38)} | ${r.gpuType.padEnd(6)} | ` +
      `${r.tier.padEnd(12)} | ${r.costPerGpuHour.toFixed(2).padStart(8)} | ${r.monthlyCost.toFixed(2).padStart(15)}`;

    if (includeCarbon) {
      line +=
        ` | ${(r.co2PerGpuHrKg || 0).toFixed(4).padStart(10)} | ` +
        `${(r.effectiveCostPerHr || 0).toFixed(4).padStart(9)} | ` +
        `${(r.renewablePct || 0).toFixed(1).padStart(8)}%`;
    }

    console.log(line);
  }

  // Best value
  const best = results[0];
  console.log(
    `\n  💡 Best value: ${best.name} at $${best.costPerGpuHour.toFixed(2)}/gpu/hr` +
    (includeCarbon ? ` (effective: $${(best.effectiveCostPerHr || best.costPerGpuHour).toFixed(4)}/gpu/hr with carbon)` : "")
  );
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const gpuType = getArg(args, "--gpu-type", "") || undefined;
  const hours = parseFloat(getArg(args, "--hours", "730"));
  const gpuCount = parseInt(getArg(args, "--gpu-count", "1"), 10);
  const includeCarbon = args.includes("--include-carbon");

  console.log("=".repeat(60));
  console.log("HarchOS Price Comparison (TypeScript)");
  console.log("=".repeat(60));

  let plans: SimPlan[] = SIMULATED_PLANS;

  // Apply GPU type filter
  if (gpuType) {
    plans = plans.filter((p) => p.gpuType === gpuType);
  }

  // Try SDK
  const apiKey = process.env.HARCHOS_API_KEY || "";
  if (apiKey) {
    try {
      const client = new HarchOSClient({ apiKey });
      const params: ListPricingPlansParams = {};
      if (gpuType) params.gpuType = gpuType as any;

      const sdkPlans = await client.pricing.listPlans(params);
      // Use SDK plans but they won't have carbon data — merge with simulated
      if (sdkPlans.length > 0) {
        console.log(`  [INFO] Retrieved ${sdkPlans.length} plans from HarchOS API`);
      }
    } catch (err) {
      console.log(`  [WARN] SDK call failed: ${err} — using simulated plans`);
    }
  }

  const results = comparePrices(plans, hours, gpuCount, includeCarbon);
  printComparison(results, includeCarbon);

  if (includeCarbon) {
    console.log("\n  🌍 Carbon-adjusted pricing includes estimated social cost of carbon");
    console.log("     at $50/tonne CO2, based on ~0.3 kWh/GPU-hour consumption.");
  }
}

function getArg(args: string[], flag: string, defaultValue: string): string {
  const idx = args.indexOf(flag);
  if (idx !== -1 && idx + 1 < args.length) return args[idx + 1];
  return defaultValue;
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: pricing/03-price-comparison/price_comparison.py
#!/usr/bin/env python3
"""Price comparison across HarchOS regions and tiers.

Compares GPU pricing plans across different regions, tiers, and GPU types
to find the most cost-effective deployment configuration. Also factors in
carbon intensity for truly optimized decision-making.

Features:
- Cross-region price comparison
- Tier-based pricing analysis (community vs enterprise vs sovereign)
- Carbon-adjusted cost scoring
- Best-value recommendation engine

Usage:
    python price_comparison.py
    python price_comparison.py --gpu-type H100
    python price_comparison.py --include-carbon
    HARCHOS_API_KEY=hsk_... python price_comparison.py
"""

import argparse
import json
import os
from typing import Dict, List, Optional

try:
    from harchos import HarchOS
    HARCHOS_AVAILABLE = True
except ImportError:
    HARCHOS_AVAILABLE = False
    print("[WARN] harchos SDK not installed — using simulated data")


# ---------------------------------------------------------------------------
# Simulated data (for local testing)
# ---------------------------------------------------------------------------

SIMULATED_PLANS = [
    {"id": "p1", "name": "H100 Enterprise (Ouarzazate)", "gpu_type": "H100",
     "price_per_gpu_hour": 2.10, "currency": "USD", "region": "morocco",
     "tier": "enterprise", "hub": "Ouarzazate", "carbon_intensity": 18,
     "renewable_pct": 97.2},
    {"id": "p2", "name": "H100 Enterprise (MAD)", "gpu_type": "H100",
     "price_per_gpu_hour": 21.00, "currency": "MAD", "region": "morocco",
     "tier": "enterprise", "hub": "Ouarzazate", "carbon_intensity": 18,
     "renewable_pct": 97.2},
    {"id": "p3", "name": "H100 Performance (Benguerir)", "gpu_type": "H100",
     "price_per_gpu_hour": 2.35, "currency": "USD", "region": "morocco",
     "tier": "enterprise", "hub": "Benguerir", "carbon_intensity": 55,
     "renewable_pct": 88.5},
    {"id": "p4", "name": "A100 Performance (Tanger)", "gpu_type": "A100",
     "price_per_gpu_hour": 1.80, "currency": "USD", "region": "morocco",
     "tier": "enterprise", "hub": "Tanger", "carbon_intensity": 95,
     "renewable_pct": 82.1},
    {"id": "p5", "name": "A100 Standard (Casablanca)", "gpu_type": "A100",
     "price_per_gpu_hour": 1.95, "currency": "USD", "region": "morocco",
     "tier": "community", "hub": "Casablanca", "carbon_intensity": 210,
     "renewable_pct": 45.0},
    {"id": "p6", "name": "L40S Enterprise (Dakhla)", "gpu_type": "L40S",
     "price_per_gpu_hour": 1.40, "currency": "USD", "region": "morocco",
     "tier": "enterprise", "hub": "Dakhla", "carbon_intensity": 32,
     "renewable_pct": 94.8},
    {"id": "p7", "name": "L40S Performance (Benguerir)", "gpu_type": "L40S",
     "price_per_gpu_hour": 1.55, "currency": "USD", "region": "morocco",
     "tier": "performance", "hub": "Benguerir", "carbon_intensity": 55,
     "renewable_pct": 88.5},
]


# ---------------------------------------------------------------------------
# Price Comparator
# ---------------------------------------------------------------------------

class PriceComparator:
    """Compares HarchOS pricing across regions, tiers, and GPU types."""

    # Carbon cost factor: higher carbon = higher effective cost
    # This represents the "hidden cost" of carbon (estimated $50/tonne CO2)
    CARBON_COST_PER_KG = 0.05  # USD per kg CO2

    def __init__(self):
        if HARCHOS_AVAILABLE:
            self.client = HarchOS(api_key=os.environ.get("HARCHOS_API_KEY", ""))
        else:
            self.client = None
        self.plans = SIMULATED_PLANS

    def fetch_plans(self, gpu_type: Optional[str] = None, tier: Optional[str] = None):
        """Fetch pricing plans, optionally filtered."""
        if self.client:
            try:
                plans = self.client.pricing.list_plans(gpu_type=gpu_type, tier=tier)
                self.plans = [
                    {
                        "id": p.id,
                        "name": p.name,
                        "gpu_type": p.gpu_type,
                        "price_per_gpu_hour": p.price_per_gpu_hour,
                        "currency": p.currency,
                        "region": p.region,
                        "tier": p.tier,
                    }
                    for p in plans
                ]
            except Exception as e:
                print(f"[WARN] SDK call failed: {e} — using simulated plans")

        # Apply filters to simulated plans too
        filtered = self.plans
        if gpu_type:
            filtered = [p for p in filtered if p["gpu_type"] == gpu_type]
        if tier:
            filtered = [p for p in filtered if p["tier"] == tier]
        return filtered

    def compare_prices(
        self,
        gpu_type: Optional[str] = None,
        tier: Optional[str] = None,
        hours: float = 730.0,
        gpu_count: int = 1,
        include_carbon: bool = False,
    ) -> List[dict]:
        """Compare prices across plans with optional carbon adjustment."""
        plans = self.fetch_plans(gpu_type=gpu_type, tier=tier)

        results = []
        for plan in plans:
            # Skip non-USD plans for fair comparison
            if plan["currency"] != "USD":
                continue

            monthly_cost = plan["price_per_gpu_hour"] * gpu_count * hours
            result = {
                **plan,
                "monthly_cost": round(monthly_cost, 2),
                "cost_per_gpu_hour": plan["price_per_gpu_hour"],
            }

            if include_carbon:
                carbon_intensity = plan.get("carbon_intensity", 100)
                # Estimated CO2 per GPU-hour: ~0.3 kWh * carbon_intensity / 1000
                co2_per_gpu_hr = 0.3 * carbon_intensity / 1000  # kg CO2
                carbon_cost_per_hr = co2_per_gpu_hr * self.CARBON_COST_PER_KG
                effective_cost_per_hr = plan["price_per_gpu_hour"] + carbon_cost_per_hr
                result["co2_per_gpu_hr_kg"] = round(co2_per_gpu_hr, 4)
                result["carbon_cost_per_hr"] = round(carbon_cost_per_hr, 4)
                result["effective_cost_per_hr"] = round(effective_cost_per_hr, 4)
                result["effective_monthly_cost"] = round(effective_cost_per_hr * gpu_count * hours, 2)
                result["carbon_intensity"] = carbon_intensity
                result["renewable_pct"] = plan.get("renewable_pct", 0)

            results.append(result)

        # Sort by effective cost (or regular cost if no carbon)
        sort_key = "effective_cost_per_hr" if include_carbon else "cost_per_gpu_hour"
        results.sort(key=lambda x: x.get(sort_key, x["cost_per_gpu_hour"]))

        return results

    def print_comparison(self, results: List[dict], include_carbon: bool = False):
        """Pretty-print the price comparison table."""
        if not results:
            print("\n  No plans found matching the criteria.")
            return

        print(f"\n  {'Rank':>4s} | {'Plan Name':38s} | {'GPU':6s} | {'Tier':12s} | "
              f"{'$/gpu/hr':>8s} | {'Monthly (1 GPU)':>15s}", end="")

        if include_carbon:
            print(f" | {'CO2 kg/hr':>10s} | {'Carbon $':>9s} | {'Eff. $/hr':>9s} | "
                  f"{'Eff. Monthly':>12s} | {'Renewable':>9s}", end="")

        print()
        print("  " + "-" * (90 if not include_carbon else 160))

        for i, r in enumerate(results, 1):
            medal = {1: "🥇", 2: "🥈", 3: "🥉"}.get(i, "  ")
            print(f"  {medal} {i:>2d} | {r['name']:38s} | {r['gpu_type']:6s} | "
                  f"{r['tier']:12s} | {r['cost_per_gpu_hour']:8.2f} | "
                  f"{r['monthly_cost']:15.2f}", end="")

            if include_carbon:
                print(f" | {r.get('co2_per_gpu_hr_kg', 0):10.4f} | "
                      f"{r.get('carbon_cost_per_hr', 0):9.4f} | "
                      f"{r.get('effective_cost_per_hr', 0):9.4f} | "
                      f"{r.get('effective_monthly_cost', 0):12.2f} | "
                      f"{r.get('renewable_pct', 0):8.1f}%", end="")

            print()

        # Best value recommendation
        if results:
            best = results[0]
            print(f"\n  💡 Best value: {best['name']} at ${best['cost_per_gpu_hour']:.2f}/gpu/hr", end="")
            if include_carbon:
                print(f" (effective: ${best.get('effective_cost_per_hr', best['cost_per_gpu_hour']):.2f}/gpu/hr "
                      f"with carbon)", end="")
            print()

    def compare_tiers(self, gpu_type: str = "H100"):
        """Compare pricing across tiers for a given GPU type."""
        print(f"\n  📊 Tier Comparison for {gpu_type}")
        print("  " + "=" * 60)

        tiers = ["community", "enterprise", "sovereign"]
        for tier in tiers:
            results = self.compare_prices(gpu_type=gpu_type, tier=tier)
            if results:
                plan = results[0]
                print(f"  {tier:12s}: ${plan['cost_per_gpu_hour']:.2f}/gpu/hr — {plan['name']}")
            else:
                print(f"  {tier:12s}: No plans available")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(description="HarchOS Price Comparison")
    parser.add_argument("--gpu-type", type=str, default=None,
                        help="Filter by GPU type (H100, A100, L40S)")
    parser.add_argument("--tier", type=str, default=None,
                        help="Filter by tier (community, enterprise, sovereign)")
    parser.add_argument("--hours", type=float, default=730.0,
                        help="Monthly hours (default: 730 = full month)")
    parser.add_argument("--gpu-count", type=int, default=1,
                        help="Number of GPUs for cost calculation")
    parser.add_argument("--include-carbon", action="store_true",
                        help="Include carbon cost in effective price")
    parser.add_argument("--compare-tiers", action="store_true",
                        help="Compare tiers for a GPU type")
    args = parser.parse_args()

    comparator = PriceComparator()

    print("=" * 60)
    print("HarchOS Price Comparison")
    print("=" * 60)

    if args.compare_tiers:
        comparator.compare_tiers(gpu_type=args.gpu_type or "H100")
        return

    results = comparator.compare_prices(
        gpu_type=args.gpu_type,
        tier=args.tier,
        hours=args.hours,
        gpu_count=args.gpu_count,
        include_carbon=args.include_carbon,
    )

    comparator.print_comparison(results, include_carbon=args.include_carbon)

    if args.include_carbon:
        print("\n  🌍 Carbon-adjusted pricing includes estimated social cost of carbon")
        print("     at $50/tonne CO2, based on ~0.3 kWh/GPU-hour consumption.")


if __name__ == "__main__":
    main()
